### ISSUE TITLE
Unable to save snapshot

### ISSUE THREAD

**Issue by macd1969** [bog-dan-ro/spectacol#153]
I searched the issues and found that others had an issue unable to save as the operation is not permitted, the solution was to become a beta tester,  I tried this and still have the same issue,  Iam using android 11,  I signed up for the beta program yesterday,  uninstalled spectacol and reinstalled from play store.  Version is showing as 1.6.05.

**Comment by YurArm** [comment]
The problem is easy to resolve. Because itis simple: the game tryes to save file with the name "filename-blblabla-date-time", 
And the problem is in "time" that looks like "19:12:05".

You should take out the illegal file characters like "colon" ( : )

**Comment by bog-dan-ro** [comment]
I'll change it to `_` in the next version. It will take some time as I need to rewrite the file handling for Android 10+.

**Comment by Dzimonchik** [comment]
we still really need a solution to this problem.  Unable to save game

**Comment by kdonath** [comment]
I have tried to change the source code and build a new apk to correct this problem but I am constantly getting the error "no rule to make target". I tried different gt versions, jdk versions, android NDK versions, but no success. I think somerhing is missing from the source.

**Comment by kdonath** [comment]
Any chance for the app to be updated?

**Comment by kdonath** [comment]
I did it! I fixed the issue. For some reason I could not recompile the source. I am missing something. BUT I edited the APK. I changed the "hh:mm:ss" to "hh_mm_ss" in the file name and now all works like a charm.

**Comment by vlig** [comment]
> I changed the "hh:mm:ss" to "hh_mm_ss" in the file name

Howto? Where?

**Comment by kdonath** [comment]
Here is what you need to do:

Tools to be used: 

On Android:
APK Extractor

On Windows PC:

APKTOOL: https://apktool.org/docs/install
https://stackoverflow.com/questions/10930331/how-to-sign-an-already-compiled-apk

*****************************************
KEYTOOL: Download JDK: https://www.oracle.com/java/technologies/downloads/

keytool path: c:\Program Files\Java\jdk-22\bin\
add to path variable

*****************************************

APKSIGNER: [https://dl.google.com/android/repository/build-tools_r30.0.1-windows.zip](https://dl.google.com/android/repository/build-tools_r30.0.1-windows.zip?authuser=0)

apksigner path: c:\Users\username\AppData\Local\Android\Sdk\build-tools\34.0.0

add to path variable

******************************************
1. Install APK Extractor to Android and extract the APK file
2. Copy the APK file (Spectacol.apk) to the Windows PC
The following steps are done on the PC

3.
Deflate APK:
apktool d Spectacol.apk

This will extract the content of the APK in a folder withh the same name.
 (APK is basically a ZIP file)

4.
Edit with Notepad++
\Spectacol\lib\arm64-v8a\[libfuse-qml_arm64-v8a.so
replace instancies of hh:mm:ss with hh_mm_ss

5.
Run CMD or Powershell in ADMIN mode

Generate Signature Key
keytool -genkey -v -keystore keystore_name.keystore -keyalg RSA -keysize 2048 -validity 10000 -alias your_alias

Password: your_keystore_password

6.
Build new APK
apktool b -o Spectacol_New.apk Spectacol

7.
Sign new APK
apksigner sign --ks-key-alias your_alias --ks keystore_name.keystore Spectacol_New.apk

8.
Copy Spectacol_New to Android.

9. On Android uninstall the original Spectacol
10. 
Allow install from local source.
On Samsung go to Settings -> Security and privacy ->  Install unknown apps -> Select My Files

11. Start My Files and browse to the Spectacol_New.apk
12. Install
13. Enjoy

**Comment by vlig** [comment]
Wow! There is the Answer! Thank you very much, I'll try it!

**Comment by qvarts** [comment]
Thank you, [kdonath](https://github.com/kdonath). This works perfect.

**Comment by Noguejordi** [comment]
Hello, good afternoon, could you please give me the modified apk? I have tried all the ways to modify it and without success, I would appreciate it if someone could give me the modified apk so I can save games, thank you very much.

**Comment by qvarts** [comment]
Hi, try this 
[Spectacol_fixed.zip](https://github.com/user-attachments/files/17093397/Spectacol_fixed.zip)

Tested on Android 13

**Comment by Noguejordi** [comment]
thank you so much

tested and working, just need to sign the apk, thanks

**Comment by Stewbie2001** [comment]
Hi, new to github but signed up to ask if there's a new updated apk available for this as I'm having the same issues with snapshot and the above updated spectacol fixed zip keeps saying it's invalid. Any help appreciated.

**Comment by kdonath** [comment]
Hi Stewbie2001,

Try this:
[Spektak2.zip](https://github.com/user-attachments/files/17939632/Spektak2.zip)

The APK file is signed and free of viruses, it works well without any issues.  But please note that I was not able to test it on all kind and versions of Android phones, so you should use it at your own responsibility.

Enjoy!

**Comment by Stewbie2001** [comment]
That's super thanks very much your a gem. I can confirm it's working on my s24 ultra just fine and I have the latest firmware on that. Thanks again bud!

**Comment by kdonath** [comment]
I am happy it works. 
The good news is that finally I managed to build a new APK from the source code. So now I can correct this problem in a more elegant way, plus I can eventually make more developments - like adding other ROMs, or support for Microdrive, etc.

**Comment by Stewbie2001** [comment]
Nice, I'm sure all us old speccy lovers will appreciate it so thanks for your hard work bud.

**Comment by Stewbie2001** [comment]
Could you increase the size of the buttons on the joysticks as I find the current ones abit on the small side.

**Comment by kdonath** [comment]
Sure, I'll take a look.

**Comment by kdonath** [comment]
[Spectacol_DK9.zip](https://github.com/user-attachments/files/18134776/Spectacol_DK9.zip)
OK, so please find attached a new version.
The changes:
- The onscreen joystick buttons are bigger - I've tested on 3 different phones, looks good to me, but please test.
- I've changed the onscreen keyboard Sticky CS/SS function. Now this works similarly to the Samsung keyboard, cycling through one time sticky - persistent sticky - released.

If you installed my previous mod (Spektak2), you need to delete if first and then install this version.
If you have the store version, this one will update the store version.
Also, once you installed it from here, you may want to turn off the automatic update (for this app only) in the Play Store app, otherwise next time you update all your apps it will try to overwrite this one with the original Store version, but will fail, because the Store version has a lower versioncode than this, and it will always give an error. If you turn off automatic update (for this app only) you avoid the error message.

**Comment by Stewbie2001** [comment]
Yup there definitely bigger, Do you think you could make them half there current size or maybe double the size they were originally.  Sorry to be a pain but thanks for the work your doing to improve this emulator as I think it's the best one available on smart phones.

**Comment by kdonath** [comment]
[Spectacol_DK9.zip](https://github.com/user-attachments/files/18137148/Spectacol_DK9.zip)
Please try this. Originally the buttons were 6mm diameter, now they are 12mm.

**Comment by Noguejordi** [comment]
On a Xiaomi Redmi note 11 pro 5g the buttons look perfect now, did you say you could put roms? Will it also be possible to search the folders from the external sd? Thanks for your work

**Comment by Stewbie2001** [comment]
Yup that's better with the size, just for reference I'm using a samsung galaxy s24

**Comment by kdonath** [comment]
[Spectacol_DK13.zip](https://github.com/user-attachments/files/20617911/Spectacol_DK13.zip)
Hello,

I have a new, improved version, if anyone interested.

Here are all the updates:

- Snapshot (szx, tzx) file name corrected. Timestamp changed to hh_mm_ss - instead of hh:mm:ss because ":" is not supported in file names on Android.
- Snapshot support: save snapshot as ..
- Onscreen keyboard sticky CS/SS cycling through one time sticky - persistent sticky - released.
- Onscreen keyboard position adjusted to better fit folding devices in tablet mode.
- Onscreen joystick buttons made bigger: 10 mm
- Microdrive support: open, save, saves as, eject functions.
- ZX Printer support: LPRINT, LLIST and COPY output to txt and pbm file.
- Package name changed, so Playstore won't try to update it anymore.

You need to delete the previous version and do a clean install.

**Comment by kdonath** [comment]
Updated version:
In Options -> General the Emulation Speed now can be directly edited - not only increased or decreased, but the actual number can be set. This is useful when you write a complex Basic program, which runs slow. While writing the program you can use 100% speed, and then run it at 900% if you wish so.

[Spectacol_DK13_3.zip](https://github.com/user-attachments/files/21184138/Spectacol_DK13_3.zip)

**Comment by Noguejordi** [comment]
Thank you for your work

**Comment by Stewbie2001** [comment]
Thanks matey for constantly improving the app.

**Comment by sc280174-droid** [comment]
Thank you so much!! I' running Spetacol on Amiga and your saving optrion fix made the apk perfect!!!

**Comment by kdonath** [comment]
Updated version: I have replaced two machine types, which I consider useless, with ROMs that are more useful.
- 16K Spectrum, in my view this makes no sense in the emulator. The new ROM is GOSH Wonderful ROM
- 48k NTSC version, also useles in my opinion. The new ROM is Pretty Basic

Descriptions of these ROMs are here: https://speccy4ever.speccy.org/_CMS.htm


The new APK:


[Spectacol_DK14_4.zip](https://github.com/user-attachments/files/24419043/Spectacol_DK14_4.zip)
